src/mesh/coordinates.cxx

```diff
@@ -249,57 +268,165 @@ Coordinates::Coordinates(Mesh *mesh, const CELL_LOC loc, const Coordinates* coor
+      break;
+    }
+  case CELL_ZLOW: {
+      // geometrical quantities are Field2D, so CELL_ZLOW version is the same
+      // as CELL_CENTRE
+      suffix = "";
+      break;
+    }
+  default: {
+      throw BoutException("Incorrect location passed to "
+          "Coordinates(Mesh*,const CELL_LOC,const Coordinates*) constructor.");
+    }
+  }
 
   nz = mesh->LocalNz;
 
   dz = coords_in->dz;
 
-  // Diagonal components of metric tensor g^{ij}
-  g11 = interpolateAndNeumann(coords_in->g11, location);
-  g22 = interpolateAndNeumann(coords_in->g22, location);
-  g33 = interpolateAndNeumann(coords_in->g33, location);
+  if (!force_interpolate_from_centre && mesh->sourceHasVar("dx"+suffix)) {
+
+    checkStaggeredGet(mesh, "dx", suffix);
+    if (mesh->get(dx, "dx"+suffix)) {
+      output_warn.write(
+          "\tWARNING: differencing quantity 'dx%s' not found. Set to 1.0\n", suffix.c_str());
+      dx = 1.0;
+    }
+    dx.setLocation(location);
 
-  // Off-diagonal elements.
-  g12 = interpolateAndNeumann(coords_in->g12, location);
-  g13 = interpolateAndNeumann(coords_in->g13, location);
-  g23 = interpolateAndNeumann(coords_in->g23, location);
+    if (mesh->periodicX) {
+      mesh->communicate(dx);
+    }
+
+    checkStaggeredGet(mesh, "dy", suffix);
+    if (mesh->get(dy, "dy"+suffix)) {
+      output_warn.write(
+          "\tWARNING: differencing quantity 'dy%s' not found. Set to 1.0\n", suffix.c_str());
+      dy = 1.0;
+    }
+    dy.setLocation(location);
+
+    // grid data source has staggered fields, so read instead of interpolating
+    // Diagonal components of metric tensor g^{ij} (default to 1)
+    getAtLoc(mesh, g11, "g11", suffix, location, 1.0);
+    getAtLoc(mesh, g22, "g22", suffix, location, 1.0);
+    getAtLoc(mesh, g33, "g33", suffix, location, 1.0);
+    getAtLoc(mesh, g12, "g12", suffix, location, 0.0);
+    getAtLoc(mesh, g13, "g13", suffix, location, 0.0);
+    getAtLoc(mesh, g23, "g23", suffix, location, 0.0);
+
+    /// Find covariant metric components
+    // Check if any of the components are present
+    if (mesh->sourceHasVar("g_11"+suffix) or mesh->sourceHasVar("g_22"+suffix) or
+        mesh->sourceHasVar("g_33"+suffix) or mesh->sourceHasVar("g_12"+suffix) or
+        mesh->sourceHasVar("g_13"+suffix) or mesh->sourceHasVar("g_23"+suffix)) {
+      // Check that all components are present
+      if (mesh->sourceHasVar("g_11"+suffix) and mesh->sourceHasVar("g_22"+suffix) and
+          mesh->sourceHasVar("g_33"+suffix) and mesh->sourceHasVar("g_12"+suffix) and
+          mesh->sourceHasVar("g_13"+suffix) and mesh->sourceHasVar("g_23"+suffix)) {
+
+        getAtLoc(mesh, g_11, "g_11", suffix, location);
+        getAtLoc(mesh, g_22, "g_22", suffix, location);
+        getAtLoc(mesh, g_33, "g_33", suffix, location);
+        getAtLoc(mesh, g_12, "g_12", suffix, location);
+        getAtLoc(mesh, g_13, "g_13", suffix, location);
+        getAtLoc(mesh, g_23, "g_23", suffix, location);
```

Comment:
Plausible `std::for_each` or `std::transform` can simplify this as well.

Reply:
This is doable with `std::transform`, but it doesn't save any lines of code, and I find the `std::transform` version harder to read (might be just not enough familiarity with the std library though).
Original version:
```
        getAtLoc(mesh, g_11, "g_11", suffix, location);
        getAtLoc(mesh, g_22, "g_22", suffix, location);
        getAtLoc(mesh, g_33, "g_33", suffix, location);
        getAtLoc(mesh, g_12, "g_12", suffix, location);
        getAtLoc(mesh, g_13, "g_13", suffix, location);
        getAtLoc(mesh, g_23, "g_23", suffix, location);
```
New version (with `getAtLoc` modified to return the `Field2D` rather than taking by reference as an argument):
```
        auto covariant_components = {g_11, g_22, g_33, g_12, g_13, g_23}; 
        auto getatloc_component = [&mesh, &suffix, &location] (const std::string name) {
          return getAtLoc(mesh, name, suffix, location);
        }
        std::transform(begin(covariant_component_names),
                       end(covariant_component_names),
                       begin(covariant_components), getatloc_component);
```
It would make more sense to use `std::transform` if we could also use `getatloc_component` for the contravariant components above, but the contravariant components need default values (1.0 for diagonal and 0.0 for off-diagonal components), which would add another layer of complication.

Reply:
Fair enough, if it was cleaner with the contravariant components as well, that it would make sense to use it. (There's also the complication of getting the right container to store the metric fields. Most of the standard containers don't store references, so they'd have to be pointers.)

It's always worth checking if there's a [standard algorithm](https://en.cppreference.com/w/cpp/algorithm) that can do the job!